### ISSUE TITLE
Duplicate project functionality between django-axes and django-defender

### ISSUE THREAD

**Issue by aleksihakli** [jazzband/django-defender#121]
I came across django-defender and thought to mention that the project has a _lot_ of duplicate functionality with django-axes (https://github.com/jazzband/django-axes).

Many of the issues in this project and django-axes are duplicate and I'm wondering if we should look into pooling our efforts and combining the projects into one code base that combines the best of all the applications?

EDIT: Removed the unrelated reference to django-watchman.

**Comment by kencochrane** [comment]
@aleksihakli I'm all for pooling resources where they make sense. What do you have in mind?

I originally wrote Defender because I was using axes and it couldn't scale to meet our needs with the current design, and it was easier to build out what we needed from scratch then try to make a major refactor on axes. The big difference is we use redis instead of the database to store state, which greatly improves the speed of the checks on login. See the README for more info.

I am not familiar with django-watchman, but looking at the link you provided, it doesn't look like there is much overlap with axes, and defender, maybe I'm missing something?

**Comment by kencochrane** [comment]
@aleksihakli like I said before, I'm open to combining projects and pooling resources, as long as it makes sense. Let's use this ticket to discuss possible options, and see where they lead us.

**Comment by joshua-s** [comment]
I think this would be good to revisit

**Comment by joshua-s** [comment]
@aleksihakli I really like your ideas around the authentication backends. For my projects, this has been the most significant difference between django-defender and django-axes. The big benefit of merging the two projects would be to expand the flexibility of the core.

**Comment by aleksihakli** [comment]
I actually just prototyped a custom authentication backend for blocking unwanted logins combined with a middleware and signal handlers for pretty error templating yesterday and managed to replicate the current functionality of django-axes 4.5 with 100% of the tests passing with zero monkey-patching.

I am going to further research django-allauth and DRF or Django REST Framework to see if they support the same approach as well, as those are some of the painpoints our current users have had.

So far the new approach seems promising, but requires some fitting together for the django-axes and django-defender.

One major issue with refactoring the project is backwards compatibility, which would be nice to keep.

**Comment by kencochrane** [comment]
I’m fine with merging the projects if it makes sense and it has the same performance as the current project. The main reason why I built defender was because axes was too slow for my requirements.

I personally don’t have time to work on any changes, so i’m not sure how much I can contribute. But i’ll try and help where I can.

**Comment by camilonova** [comment]
Lets to this. Makes total sense. I would like to see a Redis backend on axes too.

**Comment by aleksihakli** [comment]
First version of pluggable signal handlers implemented here: https://github.com/jazzband/django-axes/pull/397/commits/f4b7494e0a23cfb031ee7e0ec2e251b158d982bb

We can improve the base implementation and figure out what kind of an API interface should ideally be used for the pluggable signal handlers, but this PoC demonstrates pluggable design and I think it does it in a readable way.

**Comment by aleksihakli** [comment]
And since it might not be obvious, the other parts such as middleware and authentication backends are also pluggable, and all of the base implementations can now be parametrized and swapped. The only constraints are the APIs Django offers for middleware and authentication backends and the signal handler API we have implemented ourselves.

**Comment by aleksihakli** [comment]
@kencochrane @joshua-s @camilonova Axes has received the planned upgrades and the project layout and tests have been much improved. Could you take a look at the project and the current state of the authentication stack and give your thoughts?

**Comment by joshua-s** [comment]
@aleksihakli sorry about the delay (been alternating between sickness and heavy workloads). I hope to check this out within the next week.

**Comment by aleksihakli** [comment]
@joshua-s and others, just as a heads up, I'm keeping the PR https://github.com/jazzband/django-axes/pull/402 open for a few more days for commenting and ideas, and then merging it upstream if nobody opposes the current design. 

There are a few things that I'd like to improve in the handler and proxy layers, but as the saying goes, perfect is the enemy of good. 

I think we should try and focus the efforts on trying out and implementing the cache handler in Axes next so that we get some real experiences on extending and using the handler and the internal APIs.

**Comment by claytondaley** [comment]
FWIW we originally passed on defender/axes in favor of (the long abandoned) [django-lockount](http://pypi.python.org/pypi/django-lockout/0.1.0) because it didn't require (slower) DB calls or Redis out-of-the box.  We dedicate VMs to clients so multi-server scaling wasn't a constraint for us.

I certainly understand the value of Redis/DB backends, but I'm a huge fan of no-infrastructure, zero config options.  In this case, a zero-config in-memory option would be the simplest default.  Works grest in development and requires no special deployment effort.  If a dev actually needs Redis (or another multi-server scaling option), they should also have the expertise to switch to the appropriate backend plugin.

I think it makes sense to include a DB-backed option since it requires the least infrastructure, but it'd be nice if it was found in a self-contained package (e.g. `django-axes-backend-db`).  That way it won't create DB tables for other options (in-mem, redis, etc.).

**Comment by aleksihakli** [comment]
Hey @claytondaley! I agree on your points. The only main problem is that the local memory cache is local process only and doesn't function correctly with multiple processes, and the file based cache does not function correctly on multiple servers. 

Both can be used with Axes, but do not provide the correct functionality on multiprocess and/or multiserver setups, whereas database and Memcached or Redis setups always work correctly.

**Comment by claytondaley** [comment]
I acknowledged these points in my response; they don't change my recommendation.  Defaults should cater to the new adopter.  At that point, you're dabbling in a bunch of projects.  If it's hard to run the defaults for a particular project in a development environment, your odds of selecting that project fall precipitously.  If the defaults don't work, you're not even sure that you *can* work around the defaults (because you're brand new).  If another option works out-of-the-box, are you going to spend time digging to (1) figure out if the package has that option and (2) get the configurations working for you.

If a developer even needs multi-process or multi-server deployments, they can address it later in the development/deployment process.  They'll be more familiar with (and committed to) the package so they'll spend the time to figure out the right configurations.  They're also in the right frame of mind to decide between minimum infrastructure (DB) and maximum performance (Redis).  An in-mem option delivers on both criteria during the development/testing phase (and production for non-scaling situations).  

Obviously a big warning in the docs is both warranted and expected, as you'd find in e.g. [Django Channels 1.0](https://channels.readthedocs.io/en/1.x/getting-started.html#first-consumers) (halfway down the linked section).

**Comment by aleksihakli** [comment]
All good points, and I think the defaults should be considered carefully.

Some of my own viewpoints:

- **all projects do require correct functionality**, which the local memory and file caches do not offer under all configurations,
- **most projects do not require maximum performance setup**, and
- **switching out handlers is easy** in the new Axes version.

I am not sure if you have checked the new Axes version 5 architecture, but it uses a database handler by default, and there is a cache handler offered as well. Changing to the cache handler requires only the following setup:

```
AXES_HANDLERS = 'axes.handlers.AxesCacheHandler`
```

The Axes cache can be easily changed by adding an `axes` key to the caches and setting it to the wanted cache. Users can set it to any cache backend that Django offers internally or via 3rd party plugins.

There are checks in place to warn the user of incorrect functionality, if they try to use caches like LocMemCache, FileBasedCache, or DummyCache. Those all can have unintended functions in that they do not track the attempts correctly with multiple processes (local process), which most Django deployments run, or multiple servers (local file), which many also use, or at all (dummy).

I think that setting the default to a setting that functions incorrectly under any circumstance is a worse option than using a slightly suboptimal choice for some deployments. This is especially true for a plugin that focuses on **security**.

**Comment by aleksihakli** [comment]
Hey @kencochrane, @joshua-s, @claytondaley, and @camilonova! 

Version 5.0 has been rolled out for django-axes with multiple improvements in terms of documentation, architecture, plugability, database and cache handler implementation. The new release is pretty fast and received optimisations to both the database and cache handlers, dropping the amount of database requests to just a few transactions per login and making the cache handler more configurable as well. 

Implementing custom handlers should be pretty easy as well, making it possible to use the django-axes as a base for extensible attempt monitoring.

I'm not sure on the next steps on reducing the delta between django-axes and django-defender, but think it would be good to revisit the idea of pooling development efforts in terms of ideas and features :)

**Comment by aleksihakli** [comment]
Revisiting the subject: Doyou think there is any sense in trying to unify the projects and pool development efforts to under Jazzband, as Defender and Axes currently have duplicate functionality?

**Comment by kencochrane** [comment]
To be honest I'm not sure. What do others think? Would that mean we shut down defender and tell people to move to axes?

**Comment by aleksihakli** [comment]
I think there are multiple approaches into effort pooling including

- building a delta and feature comparison of the projects
- focusing on a specific aspect per project (performance, customization, ...)
- pooling issue backlogs and assigning effort per specific area of interest
- pooling maintenance effort and releases with joint maintenance models
- pooling codebases and combining the "good parts" into a common core or independent project

I don't think shutting down axes or defender is good since the are both nice projects with a number of users :)

**Comment by camilonova** [comment]
@kencochrane lets move this project to jazzband

**Comment by aleksihakli** [comment]
We could work on synchronizing and coordinating feature requests, security considerations etc. with a lower bar if the projects were more more integrated. That would allow for more awareness at the least. I think that evolving the ideas and gathering knowledge is more valuable than implementing nifty lines of code under one repository oro ther, and Jazzband might be a good place for that, since it gives more possibilities and visibility, but allows you to remain in control of the releases and project maintainership.

**Comment by kencochrane** [comment]
Sorry for the delay, work has been busy. I'm OK with moving it over to jazzband. I guess we just need to go over what we need to do. I'm guessing step one would be to move the project in github, and then we can go from there?

**Comment by aleksihakli** [comment]
@kencochrane the customary process is to join Jazzband at https://jazzband.co/join (that adds you as a member to the Jazzband GitHub organization) and transfer the repo to the Jazzband GitHub organization using the standard transfer workflow.

Once the repo has been transfered a new ticket is created with a TODO list to get the repo inline with the Jazzband guidelines.

Getting help with the process can be done at https://github.com/jazzband-roadies/help/issues

**Comment by kencochrane** [comment]
OK, I have joined the jazzband github org, should I transfer the project now, or wait?

**Comment by aleksihakli** [comment]
I believe transferring is OK, and then opening a ticket on the project. The things that need to be added for smooth maintenance are RTD and PyPI access to Jazzband, I think, and those should be included in the checklist that is created after transferring the project.

The main things Jazzband does is add a custom middle step in the release process: a project packet can be first uploaded to the Jazzband system, which includes a PyPI repository, and then released from the site to the public PyPI repository at https://pypi.org. If you want to utilize that, then you can add permissions for the Jazzband user on PyPI and @jezdez can add permissions for your user on the Jazzband website. The main idea of the intermediary release system is to enable safe project releases that can be done with multiple people without actually giving them the PyPI administration permissions, so you can have people safely help with the releases etc.

**Comment by jezdez** [comment]
Hey there, founder and admin of Jazzband here, I'm glad this is discussed here, please make sure to first check if the project follows the [Jazzband guidelines](https://jazzband.co/about/guidelines) before transferring it. It's customary to create a ticket to discuss this first.

@aleksihakli Please be so kind and remember to point this out when you're suggesting to move a project to Jazzband. It's vital that project owners understand that it's *NOT* just a matter of doing releases to PyPI but comes with strings attached as described in the guidelines.

@kencochrane Please do a full review of the [Jazzband guidelines](https://jazzband.co/about/guidelines) and familiarize yourself with the [release process](https://jazzband.co/about/releases) (if you wish to use it, it's optional) and look at the [FAQ](https://jazzband.co/about/faq). Thank you!

**Comment by aleksihakli** [comment]
@jezdez sure, I'll point those pages out in the future!

**Comment by kencochrane** [comment]
@aleksihakli and @jezdez Thanks for the info, I reviewed the guidelines and I think we are a good fit. The only thing that we don't have right now is a `CONTRIBUTING.md` file which we will have to create and add. Our docs right now are just a big README file, if we need to make that more formal, that would require a little bit of work.  The release process seems fine as well.

At this point would I need to fix the outstanding issues now before we transfer, or after the transfer is complete? Should we create a ticket to discuss this before we move forward with a transfer?

**Comment by aleksihakli** [comment]
There's a [copy-paste template](https://jazzband.co/about/guidelines#contributing-guidelines) for the `CONTRIBUTING.md` and `CONTRIBUTING.rst` files, whichever you'd like to use :)

Setting up rudimentary Sphinx configuration with the [Sphinx quickstart](https://www.sphinx-doc.org/en/master/usage/quickstart.html) and [Readthedocs getting started guide](https://docs.readthedocs.io/en/stable/intro/getting-started-with-sphinx.html) isn't too much work either, and it enables generating and distributing PDF docs for users who like those. I personally think that docs can be implemented and improved gradually, but setting the docs site up before the migration is nice so that users have an existing and tested basis for that.

There's some workload in setting things up that Jazzband roadies (which consist of @jezdez) have to carry out so I think it's good to discuss the topic in a ticket in the Jazzband help repository by creating a project proposal there. Proposals also allow things to go according to the existing migration process, with people getting a say in the matter and being able to consult on things.

**Comment by kencochrane** [comment]
@aleksihakli sounds good, do you want to create the ticket on the jazzband help, or should I?

**Comment by aleksihakli** [comment]
@kencochrane I think you're the best person since you are the maintainer with all the permissions and the migrations etc. will require a few steps with those.

**Comment by kencochrane** [comment]
@aleksihakli thanks, I submitted the issue, let's see what they say.

**Comment by aleksihakli** [comment]
@kencochrane I think no one is opposing moving into Jazzband :) Let me know if you'd like any help with setting up things.

**Comment by kencochrane** [comment]
@aleksihakli I haven't seen any progress on the issue I created on the jazzband-roadies repo. Is it normal that no news is good news, or should there be a formal acceptance before I move forward?

It looks like a few things that we need to do before we can move are create some sphinx docs, setup read the docs integration and then add a contributors file. If you can help with any of those, that would be great.

**Comment by jezdez** [comment]
@kencochrane Nah, that's just me being super busy with other stuff, I've just replied there. Go go go!

**Comment by kencochrane** [comment]
@jezdez Thank you, and I understand about being super busy, so no worries there.  I just transferred the repo over to jazzband, so step 1 is done. a few more to go :)

**Comment by aleksihakli** [comment]
I'm closing this ticket as I think we have concluded that we have two projects that have the same goal with different means and they should coexist. We already share knowledge and maintenance and Jazzband enables us to work and learn on the projects together :)
